File: templates/problems/11_graphs/01_num_islands/num_islands_test.go
package num_islands

import "testing"

func TestNumIslands(t *testing.T) {
	tests := []struct {
		name string
		grid [][]byte
		want int
	}{
		{
			name: "single island",
			grid: [][]byte{
				{'1', '1', '1'},
				{'0', '1', '0'},
				{'1', '1', '1'},
			},
			want: 1,
		},
		{
			name: "three islands",
			grid: [][]byte{
				{'1', '1', '0', '0', '0'},
				{'1', '1', '0', '0', '0'},
				{'0', '0', '1', '0', '0'},
				{'0', '0', '0', '1', '1'},
			},
			want: 3,
		},
		{
			name: "all water",
			grid: [][]byte{
				{'0', '0', '0'},
				{'0', '0', '0'},
			},
			want: 0,
		},
		{
			name: "all land",
			grid: [][]byte{
				{'1', '1'},
				{'1', '1'},
			},
			want: 1,
		},
		{
			name: "diagonal islands",
			grid: [][]byte{
				{'1', '0'},
				{'0', '1'},
			},
			want: 2,
		},
		{
			name: "single cell land",
			grid: [][]byte{
				{'1'},
			},
			want: 1,
		},
		{
			name: "single cell water",
			grid: [][]byte{
				{'0'},
			},
			want: 0,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// deep copy grid since it may be modified
			g := make([][]byte, len(tt.grid))
			for i := range tt.grid {
				g[i] = make([]byte, len(tt.grid[i]))
				copy(g[i], tt.grid[i])
			}
			got := numIslands(g)
			if got != tt.want {
				t.Errorf("numIslands() = %v, want %v", got, tt.want)
			}
		})
	}
}

File: templates/problems/11_graphs/01_num_islands/typescript/num_islands.test.ts
import { describe, it, expect } from "vitest";
import { numIslands } from "./num_islands";

function deepCopy<T>(grid: T[][]): T[][] {
    return grid.map((row) => [...row]);
}

describe("numIslands", () => {
    it.each([
        {
            name: "single island",
            grid: [
                ["1", "1", "1"],
                ["0", "1", "0"],
                ["1", "1", "1"],
            ],
            want: 1,
        },
        {
            name: "three islands",
            grid: [
                ["1", "1", "0", "0", "0"],
                ["1", "1", "0", "0", "0"],
                ["0", "0", "1", "0", "0"],
                ["0", "0", "0", "1", "1"],
            ],
            want: 3,
        },
        {
            name: "all water",
            grid: [
                ["0", "0", "0"],
                ["0", "0", "0"],
            ],
            want: 0,
        },
        {
            name: "all land",
            grid: [
                ["1", "1"],
                ["1", "1"],
            ],
            want: 1,
        },
        {
            name: "diagonal islands",
            grid: [
                ["1", "0"],
                ["0", "1"],
            ],
            want: 2,
        },
        {
            name: "single cell land",
            grid: [["1"]],
            want: 1,
        },
        {
            name: "single cell water",
            grid: [["0"]],
            want: 0,
        },
    ])("case: $name", ({ grid, want }) => {
        const g = deepCopy(grid);
        const got = numIslands(g);
        expect(got).toBe(want);
    });
});
